Port remaining service worker types to the new IPC serialization format

gavin-apple opened the pull request:
#### 8d41c16efed53ebf22c873ec7b5f4721c576057d
<pre>
Port remaining service worker types to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=251583">https://bugs.webkit.org/show_bug.cgi?id=251583</a>
rdar://104958760

Reviewed by Alex Christensen.

This change includes porting the following types:
    - ServiceWorkerClientType
    - ServiceWorkerJobType
    - ServiceWorkerRegistrationState
    - ServiceWorkerState
    - ServiceWorkerClientFrameType
    - ServiceWorkerClientFrameType
    - ShouldNotifyWhenResolved
    - ServiceWorkerUpdateViaCache
    - ServiceWorkerClientData
    - ServiceWorkerClientQueryOptions
    - ServiceWorkerJobData

* Source/WebCore/workers/service/ServiceWorkerClientData.h:
(WebCore::ServiceWorkerClientData::encode const): Deleted.
(WebCore::ServiceWorkerClientData::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerClientQueryOptions.h:
(WebCore::ServiceWorkerClientQueryOptions::encode const): Deleted.
(WebCore::ServiceWorkerClientQueryOptions::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerClientType.h:
* Source/WebCore/workers/service/ServiceWorkerData.h:
(WebCore::ServiceWorkerData::encode const): Deleted.
(WebCore::ServiceWorkerData::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerJob.h:
* Source/WebCore/workers/service/ServiceWorkerJobData.cpp:
(WebCore::ServiceWorkerJobData::ServiceWorkerJobData):
(WebCore::ServiceWorkerJobData::isolatedCopy const):
(WebCore::ServiceWorkerJobData::isEquivalent const):
* Source/WebCore/workers/service/ServiceWorkerJobData.h:
(WebCore::ServiceWorkerJobData::encode const): Deleted.
(WebCore::ServiceWorkerJobData::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerJobType.h:
* Source/WebCore/workers/service/ServiceWorkerTypes.h:
* Source/WebCore/workers/service/ServiceWorkerUpdateViaCache.h:
* Source/WebCore/workers/service/server/SWServerJobQueue.cpp:
(WebCore::SWServerJobQueue::runRegisterJob):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/259889@main">https://commits.webkit.org/259889@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/433a47535000bb7c4d20db027de09a3a2b2f1b1e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15371 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39156 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115510 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175613 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110227 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6587 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115198 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112080 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12793 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95775 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94668 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27418 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8611 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28770 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9120 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5342 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14743 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48316 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6832 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10685 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/workers/service/ServiceWorkerJobData.cpp:
```diff
@@ -79,8 +94,10 @@ ServiceWorkerJobData ServiceWorkerJobData::isolatedCopy() const
-
+    if (registrationOptions) {
+        ASSERT(type == ServiceWorkerJobType::Register);
+        result.registrationOptions = registrationOptions->isolatedCopy();
```
I don't see the need to make this std::optional, but if there is such a need crossThreadCopy can be used with std::optional.
Test failures look related.

gavin-apple replied:
Test failures were due to my bad validator in the decoder of registrationOptions. That's now fixed.

The reason I made registrationOptions optional is because they're only added when we have ServiceWorkerJobType::Register. It's much easier to both ASSERT this is the case during serialization using a std::optional.

Updated to use crossThreadCopy.

achristensen07 commented on Source/WebCore/workers/service/ServiceWorkerJobData.cpp:
```diff
@@ -92,11 +111,15 @@ bool ServiceWorkerJobData::isEquivalent(const ServiceWorkerJobData& job) const
             && scriptURL == job.scriptURL
             && workerType == job.workerType
-            && registrationOptions.updateViaCache == job.registrationOptions.updateViaCache;
+            && registrationOptions->updateViaCache == job.registrationOptions->updateViaCache;
```
I don't think this change is necessary.  std::optional has an operator==

gavin-apple committed:
Port remaining service worker types to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=251583
rdar://104958760

Reviewed by Alex Christensen.

This change includes porting the following types:
    - ServiceWorkerClientType
    - ServiceWorkerJobType
    - ServiceWorkerRegistrationState
    - ServiceWorkerState
    - ServiceWorkerClientFrameType
    - ServiceWorkerClientFrameType
    - ShouldNotifyWhenResolved
    - ServiceWorkerUpdateViaCache
    - ServiceWorkerClientData
    - ServiceWorkerClientQueryOptions
    - ServiceWorkerJobData

* Source/WebCore/workers/service/ServiceWorkerClientData.h:
(WebCore::ServiceWorkerClientData::encode const): Deleted.
(WebCore::ServiceWorkerClientData::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerClientQueryOptions.h:
(WebCore::ServiceWorkerClientQueryOptions::encode const): Deleted.
(WebCore::ServiceWorkerClientQueryOptions::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerClientType.h:
* Source/WebCore/workers/service/ServiceWorkerData.h:
(WebCore::ServiceWorkerData::encode const): Deleted.
(WebCore::ServiceWorkerData::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerJob.h:
* Source/WebCore/workers/service/ServiceWorkerJobData.cpp:
(WebCore::ServiceWorkerJobData::ServiceWorkerJobData):
(WebCore::ServiceWorkerJobData::isolatedCopy const):
(WebCore::ServiceWorkerJobData::isEquivalent const):
* Source/WebCore/workers/service/ServiceWorkerJobData.h:
(WebCore::ServiceWorkerJobData::encode const): Deleted.
(WebCore::ServiceWorkerJobData::decode): Deleted.
* Source/WebCore/workers/service/ServiceWorkerJobType.h:
* Source/WebCore/workers/service/ServiceWorkerTypes.h:
* Source/WebCore/workers/service/ServiceWorkerUpdateViaCache.h:
* Source/WebCore/workers/service/server/SWServerJobQueue.cpp:
(WebCore::SWServerJobQueue::runRegisterJob):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/259889@main
